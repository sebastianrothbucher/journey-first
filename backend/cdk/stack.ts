import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { LambdaRestApi, UsagePlan } from 'aws-cdk-lib/aws-apigateway'; 
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import * as checkConcreteDef from '../lambdas/check-concrete';

export class CheckConcreteStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const checkConcreteFunction = new NodejsFunction(this, 'CheckConcreteFunction', {
      runtime: cdk.aws_lambda.Runtime.NODEJS_LATEST,
      entry: __dirname + '/../lambdas/check-concrete.ts',
      handler: 'index.handler',
      bundling: {
        bundleAwsSDK: false,
        externalModules: [
          "@aws-sdk/*",
        ],
      },
      timeout: cdk.Duration.seconds(5),
    });

    const httpApi = new LambdaRestApi(this, 'JourneyFirstAPI', {
      handler: checkConcreteFunction,
      proxy: false, // confusing - just means we define lambdas individually
    });
    const checkConcreteIntegration = httpApi.root.addResource('check-concrete');
    checkConcreteIntegration.addMethod('POST', undefined, {
      apiKeyRequired: true, // !!!
    });
    checkConcreteIntegration.addCorsPreflight({
      allowOrigins: checkConcreteDef.CORS_ALLOWED_HOSTS,
      allowMethods: checkConcreteDef.CORS_ALLOWED_METHODS,
      allowHeaders: checkConcreteDef.CORS_ALLOWED_HEADERS,
    });

    const usagePlan = new UsagePlan(this, 'UsagePlan', {
      name: 'JourneyFirstUsagePlan',
      quota: {
        limit: 100,
        period: cdk.aws_apigateway.Period.WEEK,
      },
    });
    usagePlan.addApiStage({
      stage: httpApi.deploymentStage,
    });
    const key = httpApi.addApiKey("Key");
    usagePlan.addApiKey(key);

    checkConcreteFunction.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['bedrock:InvokeModel'],
        resources: ['*'],
      })
    );
  }
}