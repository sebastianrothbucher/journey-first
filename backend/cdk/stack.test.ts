import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CheckConcreteStack } from './stack';

test('Snapshot Test', () => {
  const app = new App();
  const stack = new CheckConcreteStack(app, 'test');
  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});