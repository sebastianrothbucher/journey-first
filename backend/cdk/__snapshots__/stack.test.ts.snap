// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
{
  "Outputs": {
    "JourneyFirstAPIEndpointEF5AECA6": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "JourneyFirstAPI1C8D3DDE",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "JourneyFirstAPIDeploymentStageprod378A2FFC",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CheckConcreteFunction718848E3": {
      "DependsOn": [
        "CheckConcreteFunctionServiceRoleDefaultPolicyE750F6BE",
        "CheckConcreteFunctionServiceRole57D5C5B1",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "95125e1650baaff1b966ddc45aeabd45796680bcede98868f9e08635b06d2e72.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "CheckConcreteFunctionServiceRole57D5C5B1",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CheckConcreteFunctionServiceRole57D5C5B1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CheckConcreteFunctionServiceRoleDefaultPolicyE750F6BE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "bedrock:InvokeModel",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CheckConcreteFunctionServiceRoleDefaultPolicyE750F6BE",
        "Roles": [
          {
            "Ref": "CheckConcreteFunctionServiceRole57D5C5B1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JourneyFirstAPI1C8D3DDE": {
      "Properties": {
        "Name": "JourneyFirstAPI",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "JourneyFirstAPIAccountE1F70877": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "JourneyFirstAPI1C8D3DDE",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "JourneyFirstAPICloudWatchRole95547F0C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "JourneyFirstAPICloudWatchRole95547F0C": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "JourneyFirstAPIDeploymentAF6C7BE8d6b26e49e62447a5fde972d2252be869": {
      "DependsOn": [
        "JourneyFirstAPIcheckconcreteOPTIONSD294A4CF",
        "JourneyFirstAPIcheckconcretePOST263FF349",
        "JourneyFirstAPIcheckconcreteAB9E178B",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "JourneyFirstAPI1C8D3DDE",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "JourneyFirstAPIDeploymentStageprod378A2FFC": {
      "DependsOn": [
        "JourneyFirstAPIAccountE1F70877",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "JourneyFirstAPIDeploymentAF6C7BE8d6b26e49e62447a5fde972d2252be869",
        },
        "RestApiId": {
          "Ref": "JourneyFirstAPI1C8D3DDE",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "JourneyFirstAPIKeyC7015D08": {
      "Properties": {
        "Enabled": true,
        "StageKeys": [
          {
            "RestApiId": {
              "Ref": "JourneyFirstAPI1C8D3DDE",
            },
            "StageName": {
              "Ref": "JourneyFirstAPIDeploymentStageprod378A2FFC",
            },
          },
        ],
      },
      "Type": "AWS::ApiGateway::ApiKey",
    },
    "JourneyFirstAPIcheckconcreteAB9E178B": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "JourneyFirstAPI1C8D3DDE",
            "RootResourceId",
          ],
        },
        "PathPart": "check-concrete",
        "RestApiId": {
          "Ref": "JourneyFirstAPI1C8D3DDE",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "JourneyFirstAPIcheckconcreteOPTIONSD294A4CF": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,Authorization'",
                "method.response.header.Access-Control-Allow-Methods": "'POST,OPTIONS'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "JourneyFirstAPIcheckconcreteAB9E178B",
        },
        "RestApiId": {
          "Ref": "JourneyFirstAPI1C8D3DDE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "JourneyFirstAPIcheckconcretePOST263FF349": {
      "Properties": {
        "ApiKeyRequired": true,
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "CheckConcreteFunction718848E3",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "JourneyFirstAPIcheckconcreteAB9E178B",
        },
        "RestApiId": {
          "Ref": "JourneyFirstAPI1C8D3DDE",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "JourneyFirstAPIcheckconcretePOSTApiPermissionTesttestJourneyFirstAPID023D48FPOSTcheckconcrete930C39A2": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CheckConcreteFunction718848E3",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "JourneyFirstAPI1C8D3DDE",
              },
              "/test-invoke-stage/POST/check-concrete",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "JourneyFirstAPIcheckconcretePOSTApiPermissiontestJourneyFirstAPID023D48FPOSTcheckconcrete259E7F4C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "CheckConcreteFunction718848E3",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "JourneyFirstAPI1C8D3DDE",
              },
              "/",
              {
                "Ref": "JourneyFirstAPIDeploymentStageprod378A2FFC",
              },
              "/POST/check-concrete",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "UsagePlanC18B28F1": {
      "Properties": {
        "ApiStages": [
          {
            "ApiId": {
              "Ref": "JourneyFirstAPI1C8D3DDE",
            },
            "Stage": {
              "Ref": "JourneyFirstAPIDeploymentStageprod378A2FFC",
            },
            "Throttle": {},
          },
        ],
        "Quota": {
          "Limit": 20,
          "Period": "WEEK",
        },
        "UsagePlanName": "JourneyFirstUsagePlan",
      },
      "Type": "AWS::ApiGateway::UsagePlan",
    },
    "UsagePlanUsagePlanKeyResourcetestJourneyFirstAPIKey7239C449C0FB33F8": {
      "Properties": {
        "KeyId": {
          "Ref": "JourneyFirstAPIKeyC7015D08",
        },
        "KeyType": "API_KEY",
        "UsagePlanId": {
          "Ref": "UsagePlanC18B28F1",
        },
      },
      "Type": "AWS::ApiGateway::UsagePlanKey",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
